src/cluster.c

```diff
@@ -1009,9 +1011,11 @@ clusterNode *getNodeByQuery(client *c, struct redisCommand *cmd, robj **argv, in
         mc.cmd = cmd;
     }
 
-    int is_pubsubshard = cmd->proc == ssubscribeCommand ||
-                         cmd->proc == sunsubscribeCommand ||
-                         cmd->proc == spublishCommand;
+    uint64_t cmd_flags = getCommandFlags(c);
+    
+    /* Only valid for sharded pubsub as regular pubsub can operate on any node and bypasses this layer. */
+    int pubsubshard_included = (cmd_flags & CMD_PUBSUB) ||
+                           (c->cmd->proc == execCommand && (c->mstate.cmd_flags & CMD_PUBSUB));
```

Comment:
this also affects the PUBSUB command, including PUBSUB HELP.

right below this line, we have a loop iterating on each command (handling their keys), why can't we move the previous check we had (or better yet use doesCommandHaveChannelsWithFlags) into the loop, and then set a flag named pubsubshard_included to be used in the code below the loop.?

Reply:
PUBSUB command doesn't enter `getNodeByQuery()`.
pubsub_cmd->flags&CMD_MOVABLE_KEYS == 0 and pubsub_cmd->key_specs_num == 0
```
    !(!(c->cmd->flags&CMD_MOVABLE_KEYS) && c->cmd->key_specs_num == 0 &&
          c->cmd->proc != execCommand))
    {
        int error_code;
        clusterNode *n = getNodeByQuery(c,c->cmd,c->argv,c->argc,
                                        &c->slot,&error_code);
       ....
    }
```